Introduce Request::getRequestId()

nupplaphil opened the pull request:
Because of my three different Friendica instances, each with an nginx per Friendica instance and a global reverse proxy in the front, I had a hard time to find the right requests.

I now parsed all traefik, nginx an Friendica logs, processed them and put them into a Grafana/loki storage.

With the help of this feature, I'm now able to find requests end-2-end.

An example output of Grafana/loki with the fulltext-search "b40d6d37f7a879ebffa52bf858f0a4a1":
```
2022-12-25T20:34:06+01:00	Friendica[67]: index [INFO]: Accept header was missing, using default. {"url":"https://cloud.philipp.info/remote.php/dav/files/admin/Friendica_test//0f/76/1d0d76c56b4e54e76acc2585d77d54487ef75867dc37daf8211789b9f40d","callstack":"WebDav::get, Photo::getImageDataForPhoto, Photo::rawContent, BaseModule::run"} - {"file":"HttpClient.php","line":159,"function":"request","request-id":"b40d6d37f7a879ebffa52bf858f0a4a1","uid":"a9b442"}

2022-12-25T20:34:06+01:00	{"source": "nginx", "time": 1671996846.853, "resp_body_size": 5, "host": "friendica.philipp.info", "address": "213.47.XX.XX", "request_length": 720, "method": "GET", "uri": "/photo/contact/48/263835?ts=1671918458", "status": 200,  "user_agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:108.0) Gecko/20100101 Firefox/108.0", "resp_time": 0.233, "upstream_addr": "192.168.32.6:9000", "request_id": "b40d6d37f7a879ebffa52bf858f0a4a1"}

2022-12-25T20:34:06+01:00	{"ClientAddr":"213.47.XX.XX:57190","ClientHost":"213.47.XX.XX","ClientPort":"57190","ClientUsername":"-","DownstreamContentSize":0,"DownstreamStatus":200,"Duration":233760059,"OriginContentSize":0,"OriginDuration":233509536,"OriginStatus":200,"Overhead":250523,"RequestAddr":"friendica.philipp.info","RequestContentSize":0,"RequestCount":1785,"RequestHost":"friendica.philipp.info","RequestMethod":"GET","RequestPath":"/photo/contact/48/263835?ts=1671918458","RequestPort":"-","RequestProtocol":"HTTP/2.0","RequestScheme":"https","RetryAttempts":0,"RouterName":"friendica@docker","ServiceAddr":"172.21.0.5:80","ServiceName":"web-friendica@docker","ServiceURL":{"Scheme":"http","Opaque":"","User":null,"Host":"172.21.0.5:80","Path":"","RawPath":"","OmitHost":false,"ForceQuery":false,"RawQuery":"","Fragment":"","RawFragment":""},"StartLocal":"2022-12-25T19:34:06.619883418Z","StartUTC":"2022-12-25T19:34:06.619883418Z","TLSCipher":"TLS_AES_128_GCM_SHA256","TLSVersion":"1.3","downstream_X-Request-Id":"b40d6d37f7a879ebffa52bf858f0a4a1","entryPointName":"websecure","level":"info","msg":"","origin_X-Request-Id":"b40d6d37f7a879ebffa52bf858f0a4a1","request_User-Agent":"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:108.0) Gecko/20100101 Firefox/108.0","time":"2022-12-25T19:34:06Z"}
```

Screenshot of the new error page:
![Bildschirmfoto vom 2022-12-26 13-14-19](https://user-images.githubusercontent.com/379654/209547918-8c63ace0-a7b5-4cd5-8cf9-803b4cc41d10.png)


The next step is to add an addon "JSON Logger" to replace my "hacky" log-processor ;)

nupplaphil committed:
Adapt Logger\Introspection
- Create an interface
- Add it as constructor parameter

nupplaphil committed:
Use X-REQUEST-ID for Logging

nupplaphil committed:
adapt examples

MrPetovan commented on src/Module/HTTPException/PageNotFound.php:
```diff
@@ -77,6 +77,6 @@ public function run(array $request = []): ResponseInterface
 		]);
 
-		return parent::run($request); // TODO: Change the autogenerated stub
+		return parent::run($requestId, $request); // TODO: Change the autogenerated stub
```
```suggestion
		return parent::run($requestId, $request);
```

MrPetovan commented on src/Module/Special/HTTPException.php:
```diff
@@ -41,14 +49,15 @@ class HTTPException
 		$vars = [
 			'$title' => $e->getDescription() ?: 'Error ' . $e->getCode(),
 			'$message' => $e->getMessage() ?: $e->getExplanation(),
 			'$back' => DI::l10n()->t('Go back'),
 			'$stack_trace' => DI::l10n()->t('Stack trace:'),
```
Please remove the dependency to `DI` since you made the method dynamic.

nupplaphil replied:
I knew it *g*
(I was just a little bit lazy and thought maybe you won't catch it :D ..)

nupplaphil committed:
Adapt UserSession
- Move from App methods to UserSession methods
- Deprecate corresponding App methods

nupplaphil commented:
@MrPetovan - plz review, I'm still testing it at my node

MrPetovan commented on src/App.php:
```diff
@@ -136,64 +133,39 @@ class App
-	 *
-	 * @param int $user_id
-	 * @return void
+	 * @deprecated 2022.03
```
Yeeees!

MrPetovan commented on src/Core/Session/Model/UserSession.php:
```diff
@@ -50,6 +51,16 @@ public function getLocalUserId()
+	/** {@inheritDoc} */
+	public function getLocalUserNickname()
+	{
+		if (!empty($this->session->get('authenticated')) && !empty($this->session->get('nickname'))) {
+			return $this->session->get('nickname');
+		}
```
```suggestion
		if ($this->isAuthenticated()) {
			return $this->session->get('nickname') ?? false;
		}
```

MrPetovan commented on src/Core/Session/Model/UserSession.php:
```diff
@@ -122,6 +133,13 @@ public function isAuthenticated(): bool
+	/** {@inheritDoc} */
+	public function isSiteAdmin(): bool
+	{
+		return $this->getLocalUserId() &&
+			   User::isSiteAdmin($this->getLocalUserId());
```
```suggestion
		return User::isSiteAdmin($this->getLocalUserId());
```

MrPetovan commented on src/Module/Special/HTTPException.php:
```diff
@@ -33,27 +36,52 @@
+	/** @var string */
+	protected $requestId;
+
+	public function __construct(L10n $l10n, LoggerInterface $logger, Arguments $args, UserSession $session, Request $request, array $server = [])
```
Shouldn't this class inherit from `BaseModule`? 🤔

nupplaphil replied:
No, because it must not be used at the normal `BaseModule::run()` and not retrieved by the `App\Router`. It's directly called with an exception.

And additionally, if I do inherit it from `BaseModule`, I cannot use the `rawContent()` and `content()` calls anymore, because they're called with an `HTTPException` instead of `array $request`.

So it make things more complicated.

Another proof for me that there's a design flaw at `BaseModule`...

MrPetovan commented on tests/src/Module/Special/OptionsTest.php:
```diff
@@ -24,16 +24,27 @@
+		parent::setUp();
+
+		$this->httpExceptionMock = \Mockery::mock(HTTPException::class);
+	}
 	public function testOptionsAll()
```
```suggestion
	}

	public function testOptionsAll()
```

nupplaphil commented on src/Capabilities/ICanHandleRequests.php:
```diff
@@ -30,11 +31,12 @@
 	 * @throws HTTPException\InternalServerErrorException
 	 */
-	public function run(array $request = []): ResponseInterface;
+	public function run(ModuleHTTPException $httpException, array $request = []): ResponseInterface;
```
@MrPetovan jfyi

I'm really not happy about extending the `run()`. I think it's a design flaw and leads to a general design flaw I introduced with `BaseModule`.

I think rather than executing `run()` from inside `BaseModule`, including all the `run()` depending methods, there should be a `ModuleExecutor` class (just a working title)  and so all these dependencies in the constructor will almost vanish, so new Modules won't have to add these 7 dependencies in the constructor and the `run()` won't need to be overridden with new Modules..

--> followup

I already started, but it would extend this PR way too much

MrPetovan replied:
Would this refactoring take place before or after this PR?

nupplaphil replied:
After :-)
Step by step

MrPetovan replied:
All good, I should be notified when you open this PR for merging.

nupplaphil commented:
Found two errors and fixed them.

Now I'm ready :-)

nupplaphil committed:
Use X-REQUEST-ID for Error pages
